### PR TITLE
ARROW-6887: [Java] Create prose documentation for using ValueVectors

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-6887](https://issues.apache.org/jira/browse/ARROW-6887).
We should create documentation (in restructured text) for the library that demonstrates:

1.  Basic construction of ValueVectors.  Highlighting:
    * ValueVector lifecycle
    * Reading by rows using Readers (mentioning that it is not as efficient as direct access).
    * Populating with Writers
2.  Reading and writing IPC stream format and file formats.

### CODE REVIEW

**Comment by tianchen92** [comment]
cc @emkornfield Thanks for your careful review, I updated the document. Hmm, and the original PR was in a wrong branch(my local master), so I closed it and create this one.

ping @jacques-n , please help take a look, thanks!

**Comment by tianchen92** [comment]
The travis always fail, I am not sure if it's caused by the docs?

File: docs/source/java/vector_schema_root.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+================
+VectorSchemaRoot
+================
+A :class:`VectorSchemaRoot` is a container that can hold batches, batches flow through :class:`VectorSchemaRoot`
+as part of a pipeline. Note this is different from other implementations (i.e. in C++ and Python,
+a :class:`RecordBatch` is a collection of equal-length vector instances and was created each time for a new batch).
+
+The recommended usage for :class:`VectorSchemaRoot` is creating a single :class:`VectorSchemaRoot` based on
+ the known schema and populated data over and over into the same VectorSchemaRoot in a stream of batches rather than
+ create a new :class:`VectorSchemaRoot` instance each time (see ``Flight`` or ``ArrowFileWriter`` for better
```
Comment (emkornfield) [inline docs/source/java/vector_schema_root.rst]: can you provide specific links to the code for flight?

How does one transfer vectors from one VectorSchemaRoot to another one?  Or is one VectorSchemaRoot intented to be shared across multiple processors?

**Comment by emkornfield** [comment]
@tianchen92 from the logs:
"/home/travis/build/apache/arrow/docs/source/java/index.rst:19: WARNING: Title underline too short."

I think you need to add an equals sign to that line

File: docs/source/java/vector_schema_root.rst (hunk 1 above)
Comment (tianchen92) [reply docs/source/java/vector_schema_root.rst]: We could get a new VectorSchemaRoot instance by VectorShemaRoot.slice API(do transfer for its vectors), but I am not sure if this is the right usage considering VectorSchemaRoot != recordBatch

**Comment by tianchen92** [comment]
> @tianchen92 from the logs:
> "/home/travis/build/apache/arrow/docs/source/java/index.rst:19: WARNING: Title underline too short."
> 
> I think you need to add an equals sign to that line

Thanks, I updated and see if it could pass.

File: docs/source/java/ipc.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+===========================
+Reading/Writing IPC formats
+===========================
+Arrow defines two types of binary formats for serializing record batches:
+
+* **Streaming format**: for sending an arbitrary length sequence of record
```
Comment (liyafan82) [inline docs/source/java/ipc.rst]: arbitrary length sequence of -> arbitrary number of ?

File: docs/source/java/ipc.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  batches. The format must be processed from start to end, and does not support
+  random access
+
+* **File or Random Access format**: for serializing a fixed number of record
+  batches. Supports random access, and thus is very useful when used with
```
Comment (liyafan82) [inline docs/source/java/ipc.rst]: Supports -> It supports ?

File: docs/source/java/ipc.rst [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    List<FieldVector> vectors = Arrays.asList(bitVector, varCharVector);
+    VectorSchemaRoot root = new VectorSchemaRoot(fields, vectors);
+
+Now, we can begin writing a stream containing some number of these batches. For this we use :class:`ArrowStreamWriter`
+(DictionaryProvider used for any vectors that are dictionary encoded is optional and can be null))::
+
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    ArrowStreamWriter writer = new ArrowStreamWriter(root, /*DictionaryProvider=*/null, Channels.newChannel(out));
+
+Here we used an in-memory stream, but this could have been a socket or some other IO stream, then we can do::
```
Comment (liyafan82) [inline docs/source/java/ipc.rst]: some other IO stream, then we -> some other IO stream. Then we ...

File: docs/source/java/ipc.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      childVector2.reset();
+      ... do some populate work here, could be different for each batch
+
+      writer.writeBatch();
+    }
+    writer.end();
+
+Note since the :class:`VectorSchemaRoot` in writer is a container that can hold batches, batches flow through
+:class:`VectorSchemaRoot` as part of a pipeline, so we need to populated data if necessary before non-first call,
+otherwise several same batches will be written.
```
Comment (liyafan82) [inline docs/source/java/ipc.rst]: Do you mean "later batches will overwrite previous ones"?

File: docs/source/java/ipc.rst [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+Now the :class:`ByteArrayOutputStream` contains the complete stream which contains 5 record batches.
```
Comment (liyafan82) [inline docs/source/java/ipc.rst]: Why 5 batches? In the previous code, the loop runs 4 times.

File: docs/source/java/ipc.rst [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      Schema schema = reader.getVectorSchemaRoot().getSchema();
+      for (int i = 0; i < 5; i++) {
+        // This will be loaded with new values on every call to loadNextBatch
+        VectorSchemaRoot readBatch = reader.getVectorSchemaRoot();
+        assertTrue(reader.loadNextBatch());
+        // do something with readBatch
+      }
+
+    }
+
```
Comment (liyafan82) [inline docs/source/java/ipc.rst]: Can you please give an example when the dictionary is not null?

File: docs/source/java/vector.rst [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+===========
+ValueVector
+===========
+
+:class:`ValueVector` (which called Array in C++ implementation and the :doc:`the specification <../format/Columnar>`)
```
Comment (liyafan82) [inline docs/source/java/vector.rst]: This maybe a little nit-picking: ValueVector is an interface?

File: docs/source/java/vector.rst [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+Building ValueVector
+====================
+
+Note that the current implementation doesn't enforce the rule that Arrow objects are immutable.
+:class:`ValueVector` instances could be created directly by using new keyword, there are
+set/setSafe APIs and concrete subclasses of FieldWriter for populating values.
+
+For example, the code below shows how to build a :class:`BigIntVector`, in this case, we build a
+vector of the range 1 to 8 where the element that should hold the value 4 is nulled::
```
Comment (liyafan82) [inline docs/source/java/vector.rst]: 1 to 8 ->0 to 7?
value 4 -> the fourth value?

File: docs/source/java/vector.rst [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+   vector.set(2, 3);
+   vector.setNull(3);
+   vector.set(4, 5);
+   vector.set(5, 6);
+   vector.set(6, 7);
+   vector.set(7, 8);
+   vector.setValueCount(8); // this will finalizes the vector by convention.
+
+The :class:`BigIntVector` holds two ArrowBufs. The first buffer holds the null bitmap, which consists
+here of a single byte with the bits 0|0|0|0|1|0|0|0. The second buffer contains all the above values.
```
Comment (liyafan82) [inline docs/source/java/vector.rst]: The bits in the bit map is incorrect. Please note that the bit is 1 if the value is non-null.

File: docs/source/java/vector.rst [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+Here is how to build a vector using writer::
+
+   BigIntVector vector = new BigIntVector("vector", allocator);
+   BigIntWriter writer = new BigIntWriterImpl(vector);
+   writer.setPosition(0);
+   writer.writeBigInt(1);
+   writer.setPosition(1);
+   writer.writeBigInt(2);
+   writer.setPosition(2);
+   writer.writeBigInt(3);
```
Comment (liyafan82) [inline docs/source/java/vector.rst]: Also give an example for writing null?

File: docs/source/java/vector.rst [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    IntVector vector = new IntVector("intVector", allocator);
+    for (int i = 0; i < 10; i++) {
+      vector.setSafe(i, i);
+    }
+    vector.setValueCount(10);
+
+    TransferPair tp = vector.getTransferPair(allocator);
+    tp.splitAndTransfer(0, 5);
+    IntVector sliced = (IntVector) tp.getTo();
+
```
Comment (liyafan82) [inline docs/source/java/vector.rst]: Briefly describe the content of "vector" and "sliced", after calling splitAndTransfer?

File: docs/source/java/vector_schema_root.rst [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+================
+VectorSchemaRoot
+================
+A :class:`VectorSchemaRoot` is a container that can hold batches, batches flow through :class:`VectorSchemaRoot`
+as part of a pipeline. Note this is different from other implementations (i.e. in C++ and Python,
+a :class:`RecordBatch` is a collection of equal-length vector instances and was created each time for a new batch).
+
+The recommended usage for :class:`VectorSchemaRoot` is creating a single :class:`VectorSchemaRoot`
+based on the known schema and populated data over and over into the same VectorSchemaRoot in a stream
+of batches rather than create a new :class:`VectorSchemaRoot` instance each time
```
Comment (liyafan82) [inline docs/source/java/vector_schema_root.rst]: create -> creating

File: docs/source/java/vector_schema_root.rst [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+(see `Numba <https://github.com/apache/arrow/tree/master/java/flight/src/main/java/org/apache/arrow/flight>`_ or
+``ArrowFileWriter`` for better understanding). Thus at any one point a VectorSchemaRoot may have data or
+may have no data (say it was transferred downstream or not yet populated).
+
+
+Here is the example of building a :class:`VectorSchemaRoot`::
+
+    BitVector bitVector = new BitVector("boolean", allocator);
+    VarCharVector varCharVector = new VarCharVector("varchar", allocator);
```
Comment (liyafan82) [inline docs/source/java/vector_schema_root.rst]: we should call allocateNew here?

File: docs/source/java/vector_schema_root.rst [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    bitVector.setValueCount(10);
+    varCharVector.setValueCount(10);
+
+    List<Field> fields = Arrays.asList(bitVector.getField(), varCharVector.getField());
+    List<FieldVector> vectors = Arrays.asList(bitVector, varCharVector);
+    VectorSchemaRoot vectorSchemaRoot = new VectorSchemaRoot(fields, vectors);
+
+The vectors within a :class:`VectorSchemaRoot` could be loaded/unloaded via :class:`VectorLoader` and :class:`VectorUnloader`.
+:class:`VectorLoader` and :class:`VectorUnloader` handles converting between :class:`VectorSchemaRoot` and :class:`ArrowRecordBatch`(
+representation of an RecordBatch :doc:`IPC <../format/IPC.rst>` message). Examples as below::
```
Comment (liyafan82) [inline docs/source/java/vector_schema_root.rst]: an -> a

File: docs/source/java/vector_schema_root.rst [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    ArrowRecordBatch recordBatch = unloader.getRecordBatch();
+
+    // create a VectorSchemaRoot root2 and load the recordBatch
+    VectorSchemaRoot root2 = VectorSchemaRoot.create(root1.getSchema(), allocator);
+    VectorLoader loader = new VectorLoader(root2);
+    loader.load(recordBatch);
+
+A new :class:`VectorSchemaRoot` could be sliced from an existing instance with zero-copy::
+
+    // 0 indicates start index and 5 indicated length.
```
Comment (liyafan82) [inline docs/source/java/vector_schema_root.rst]: 0 indicates start index and 5 indicated length ->
0 indicates start index (inclusive) and 5 indicated length (exclusive)

File: docs/source/java/ipc.rst (hunk 6 above)
Comment (tianchen92) [reply docs/source/java/ipc.rst]: Because we write a batch already before the loop.

File: docs/source/java/vector.rst (hunk 11 above)
Comment (tianchen92) [reply docs/source/java/vector.rst]: The fourth value is still null like the example with setAPI, I updated the comments.

**Comment by tianchen92** [comment]
@liyafan82 Thanks a lot for your comments/corrections, I updated this PR.

File: docs/source/java/ipc.rst [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+Note since the :class:`VectorSchemaRoot` in writer is a container that can hold batches, batches flow through
+:class:`VectorSchemaRoot` as part of a pipeline, so we need to populate data before `writeBatch` so that later batches
+could overwrite previous ones.
+
+Now the :class:`ByteArrayOutputStream` contains the complete stream which contains 5 record batches.
+We can read such a stream with :class:`ArrowStreamReader`, note that :class:`VectorSchemaRoot` within
+reader will be loaded with new values on every call to :class:`loadNextBatch()`::
+
+    try (ArrowStreamReader reader = new ArrowStreamReader(new ByteArrayInputStream(out.toByteArray()), allocator)) {
```
Comment (emkornfield) [inline docs/source/java/ipc.rst]: does this close the associated VectorSchemaRoot as well?

**Comment by emkornfield** [comment]
this seems like a good start.  @jacques-n do you have time to take a quick look to make sure it aligns with your understanding of usage?  Otherwise I'll merge in the next couple of days.

File: docs/source/java/ipc.rst (hunk 17 above)
Comment (tianchen92) [reply docs/source/java/ipc.rst]: yes, reader#close will do this

**Comment by emkornfield** [comment]
+1, thank you this is a good start to making the java library more accessible.

**Pull request closed without merging**